web/src/engine/package-cache/src/modelCache.ts

```diff
@@ -0,0 +1,66 @@
+import { ModelSpec } from "@keymanapp/input-processor";
+
+export default class ModelManager {
+  // Tracks registered models by ID.
+  private registeredModels: {[id: string]: ModelSpec} = {};
+
+  // Allows for easy model lookup by language code; useful when switching keyboards.
+  languageModelMap: {[language:string]: ModelSpec} = {};
+
+  modelForLanguage(lgCode: string) {
+    return this.languageModelMap[lgCode];
+  }
+
+  // Accessible publicly as keyman.modelManager.register(model: ModelSpec)
+  register(model: ModelSpec): void {
+    // Forcibly lowercase the model ID before proceeding.
+    model.id = model.id.toLowerCase();
+
+    if(JSON.stringify(model) == JSON.stringify(this.registeredModels[model.id])) {
+      // We are already registered, let's not go through and re-register
+      // because we'll already have the correct model active
+      return;
+    }
+    this.registeredModels[model.id] = model;
+
+    // Register the model for each targeted language code variant.
+    let mm = this;
+    model.languages.forEach(function(code: string) {
+      // Prevent null / undefined codes; they're invalid.
+      if(!code) {
+        console.warn("Null / undefined language codes are not permitted for registration.");
+        return;
+      }
+
+      mm.languageModelMap[code] = model;
+    });
+  }
+
+  deregister(modelId: string): ModelSpec {
```

Comment:
Probably should be `unregister` because we use that terminology a bit with keyboards...

Reply:
Can do.

On a whim, I decided to check on the difference between "unregister" and "deregister":

https://english.stackexchange.com/questions/25931/unregister-vs-deregister

... yeah, let's leave that can of worms unopened.